### PR TITLE
Swap_letters review

### LINKED ISSUE #15: Challenge_Easy_swAp_LetterS
The function convert all lowercase letters to uppercase letters and vice versa.

Input: string
Return: the modified string 

Sample input:
"Hello World"

Output:
"hELLO wORLD"

Refer to this short tutorial for help. 
https://www.geeksforgeeks.org/isupper-islower-lower-upper-python-applications/

### PR DESCRIPTION
**Pull request by TibyanKhalid**
---
name: solution review 
about: A template PR for code review with a checklist
---

The function convert all lowercase letters to uppercase letters and vice versa.

Input: string
Return: the modified string

Sample input:
"Hello World"

Output:
"hELLO wORLD"

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
  - _It's ok to have extra helper functions if necessary, like with mergesort_
- [x] The function has correct type annotations
- [x] The function is not called at the top level of the function file 
  - _Recursive solutions **can** call the function from **inside** the function body_

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in any docstrings or tests
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

File: solutions/Swap_letters.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
+"""
+Created on 08/01/2024
+
+@author: Tibyan Khalid
+"""
```
Comment (Frank2446-dotcom) [inline solutions/Swap_letters.py:7]: The docstring is good for keeping track of the file's creation and author. However, it would be more helpful if you included a brief description of the file's purpose or functionality here. For example:

"""This file contains the function Swap_letters that swaps the case of each letter in a string."""

File: solutions/Swap_letters.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+    >>> Swap_letters("hello")
+    'HELLO'
+
+    >>> Swap_letters("HELLO")
+    'hello'
+
+    >>> Swap_letters("HeLlO")
+    'hElLo'
+    """
```
Comment (Frank2446-dotcom) [inline solutions/Swap_letters.py:26]: The docstring is clear and provides example usage (which is good for doctests). However:

Add edge cases in the doctests, such as empty strings, strings with digits, or special characters.

File: solutions/Swap_letters.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    if not isinstance(string, str):
+        "Make sure the expected input is a string"
```
Comment (Frank2446-dotcom) [inline solutions/Swap_letters.py:28]: The string literal "Make sure the expected input is a string" doesn’t do anything in the code. It should either be removed or converted to a comment:

# Make sure the expected input is a string

File: solutions/Swap_letters.py [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+    changed_string = ""
+    for char in string:
+        if char.islower():
+            changed_string = changed_string + char.upper()
+        elif char.isupper():
+            changed_string = changed_string + char.lower()
+        else:
+            changed_string = changed_string + char
+    return changed_string
```
Comment (Frank2446-dotcom) [inline solutions/Swap_letters.py:39]: The function implementation is clear, but you can simplify the concatenation in the loop:
changed_string += char.upper() if char.islower() else char.lower() if char.isupper() else char

Using += is cleaner and avoids multiple concatenations.

File: solutions/tests/test_Swap_letters.py [hunk 5]
```diff
@@ -0,0 +1,38 @@
+"""
+Created on 08/01/2024
+
+@author: Tibyan Khalid
+"""
```
Comment (Frank2446-dotcom) [inline solutions/tests/test_Swap_letters.py:7]: Same as the previous file—consider adding a description of what the test file is testing

File: solutions/tests/test_Swap_letters.py [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+
+import unittest
+
+from ..Swap_letters import Swap_letters
+
+
+class TestSwapletters(unittest.TestCase):
+    """Unittests for the Swap_letters function"""
```
Comment (Frank2446-dotcom) [inline solutions/tests/test_Swap_letters.py]: The class name TestSwapletters should be in PascalCase, which is correct. However, the name should be more descriptive, such as TestSwapLettersFunctionality to better describe what exactly is being tested.

The docstring for the class is good, but could be more specific. For example:

"""Test cases for validating the behavior of the Swap_letters function, including edge cases."""

File: solutions/tests/test_Swap_letters.py [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+
+    def test_lowercase_all(self):
+        "Testing from lowercase to uppercase"
+        self.assertEqual(Swap_letters("tibyan"), "TIBYAN")
```
Comment (Frank2446-dotcom) [inline solutions/tests/test_Swap_letters.py:20]: It’s a good idea to add edge case tests, such as testing an empty string or special characters.
You could consider adding more boundary tests for edge cases (empty strings, non-letters).

**Review by Frank2446-dotcom — CHANGES_REQUESTED** [review]

File: solutions/Swap_letters.py (hunk 1 above)
Comment (TibyanKhalid) [reply solutions/Swap_letters.py:7]: Good idea yeah okay I'll do that now

File: solutions/Swap_letters.py (hunk 2 above)
Comment (TibyanKhalid) [reply solutions/Swap_letters.py:26]: I've added the edge cases tests in the tests file already

File: solutions/Swap_letters.py (hunk 4 above)
Comment (TibyanKhalid) [reply solutions/Swap_letters.py:39]: Am I the only one who doesn't like " += " haha, but yeah I think it definitely is clearer I'll change it now

File: solutions/tests/test_Swap_letters.py (hunk 5 above)
Comment (TibyanKhalid) [reply solutions/tests/test_Swap_letters.py:7]: Okay sure

File: solutions/Swap_letters.py (hunk 3 above)
Comment (TibyanKhalid) [reply solutions/Swap_letters.py:28]: It's just for documenting purposes, it acts as a comment

File: solutions/tests/test_Swap_letters.py (hunk 6 above)
Comment (TibyanKhalid) [reply solutions/tests/test_Swap_letters.py]: Noted I'll make it more descriptive right now

File: solutions/tests/test_Swap_letters.py (hunk 7 above)
Comment (TibyanKhalid) [reply solutions/tests/test_Swap_letters.py:20]: I thought I added them omg, okay I'll add them right nowww

**Review by Frank2446-dotcom — APPROVED** [review]
Nice Job
